### PR TITLE
tests: randomized order for PT & Fabric

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

random test execution to prevent accidental test depenency

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Comment by Borda** [comment]
## Pytorch

```
FAILED core/test_metric_result_integration.py::test_metric_result_respects_dtype[floating_dtype0] - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.
The list of emitted warnings is: [].
```

## Fabric

```
FAILED loggers/test_csv.py::test_file_logger_no_name[] - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-runner/pytest-0/test_file_logger_no_name__0/version_0'
FAILED loggers/test_csv.py::test_file_logger_log_hyperparams - Failed: DID NOT RAISE <class 'NotImplementedError'>
FAILED loggers/test_csv.py::test_file_logger_log_metrics[10] - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-runner/pytest-0/test_file_logger_log_metrics_10/lightning_logs/version_0/metrics.csv'
FAILED loggers/test_csv.py::test_file_logger_named_version - AssertionError: assert [] == ['2020-02-05-162402']
  Right contains one more item: '2020-02-05-162402'
  Full diff:
  - ['2020-02-05-162402']
  + []
FAILED loggers/test_csv.py::test_flush_n_steps - AssertionError: Expected 'mock' to have been called once. Called 0 times.
FAILED loggers/test_csv.py::test_file_logger_no_name[None] - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-runner/pytest-0/test_file_logger_no_name_None_0/version_0'
FAILED loggers/test_csv.py::test_file_logger_log_metrics[None] - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-runner/pytest-0/test_file_logger_log_metrics_N0/lightning_logs/version_0/metrics.csv'
FAILED test_wrappers.py::test_fabric_module_method_lookup - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.
The list of emitted warnings is: [].
```

## App

```
FAILED tests_app/utilities/test_network.py::test_lightning_client_retry_enabled - Failed: Timeout >300.0s
FAILED tests_app/utilities/test_logs_socket_api.py::test_cluster_logs_socket_api - Failed: Timeout >300.0s
FAILED tests_app/core/test_lightning_api.py::test_app_state_api - Failed: Timeout >300.0s
FAILED tests_app/core/test_lightning_api.py::test_app_state_api_with_flows - Failed: Timeout >300.0s
```

File: requirements/app/test.txt [hunk 1]
```diff
@@ -4,11 +4,13 @@ pytest-timeout ==2.1.0
 pytest-cov ==4.0.0
 pytest-doctestplus >=0.9.0
 pytest-asyncio ==0.20.3
+pytest-random-order <=1.1.0
```
Comment (carmocca) [inline requirements/app/test.txt:7]: Tesst requirements can and should pin to the latest version

Instead of `<=`

**Review by carmocca — COMMENTED** [review]
Pushed a commit that might fix the Trainer test failure

Overall the idea of running a randomized order sounds good to me

File: requirements/app/test.txt (hunk 1 above)
Comment (Borda) [reply requirements/app/test.txt:7]:
```suggestion
pytest-random-order ==1.1.0
```

File: requirements/fabric/test.txt [hunk 2]
```diff
@@ -1,6 +1,7 @@
-coverage==6.5.0
-pytest==7.2.2
-pytest-cov==4.0.0
-pytest-rerunfailures==10.3
-click==8.1.3
+coverage ==6.5.0
+pytest ==7.2.2
+pytest-cov ==4.0.0
+pytest-rerunfailures ==10.3
+pytest-random-order <=1.1.0
```
Comment (Borda) [inline requirements/fabric/test.txt:5]:
```suggestion
pytest-random-order ==1.1.0
```

File: requirements/pytorch/test.txt [hunk 3]
```diff
@@ -1,18 +1,19 @@
...
-pytest==7.2.2
-pytest-cov==4.0.0
-pytest-forked==1.4.0
-pytest-rerunfailures==10.3
+coverage ==6.5.0
+pytest ==7.2.2
+pytest-cov ==4.0.0
+pytest-forked ==1.4.0
+pytest-rerunfailures ==10.3
+pytest-random-order <=1.1.0
```
Comment (Borda) [inline requirements/pytorch/test.txt:6]:
```suggestion
pytest-random-order ==1.1.0
```

File: requirements/app/test.txt [hunk 4]
```diff
@@ -4,11 +4,13 @@ pytest-timeout ==2.1.0
 pytest-cov ==4.0.0
 pytest-doctestplus >=0.9.0
 pytest-asyncio ==0.20.3
+pytest-random-order ==1.1.0
 pytest-rerunfailures <=11.1.2
+
 playwright ==1.30.0
-httpx
+httpx <=0.24.0
```
Comment (carmocca) [inline requirements/app/test.txt:11]: My previous comment also applies to all other `<=` or `>=` in the different `test.txt` files. Maybe you prefer to do it in a separate PR. Up to you

Comment (Borda) [reply requirements/app/test.txt:11]:
```suggestion
httpx ==0.24.0
```

**Comment by Borda** [comment]
Signing now to @awaelchli and @ethanwharris if they have validated the remaining failing tests, but no pressure :rabbit:

File: requirements/pytorch/test.txt [hunk 5]
```diff
@@ -1,18 +1,19 @@
...
 cloudpickle >=1.3, <2.3.0
-scikit-learn>0.22.1, <1.2.1
-onnx<1.14.0
-onnxruntime<1.14.0
-psutil<5.9.5 # for `DeviceStatsMonitor`
+scikit-learn >0.22.1, <1.2.1
+onnx <1.14.0
+onnxruntime <1.14.0
+psutil <5.9.5 # for `DeviceStatsMonitor`
 pandas>1.0, <2.0.2  # needed in benchmarks
```
Comment (Borda) [inline requirements/pytorch/test.txt:14]:
```suggestion
pandas >1.0, <2.0.2  # needed in benchmarks
```

**Comment by Borda** [comment]
seems Fabric is solved, just App holds

File: .github/workflows/ci-tests-app.yml [hunk 6]
```diff
@@ -124,8 +124,7 @@ jobs:
         AWS_DEFAULT_REGION: us-east-1
       run: |
         python -m coverage run --source ${{ env.COVERAGE_SCOPE }} \
-          -m pytest -m "not cloud" tests_app \
-          --timeout=300 -vvvv --durations=50
+          -m pytest -m "not cloud" tests_app --timeout=300 -vvvv --durations=50 --random-order-seed=42
```
Comment (carmocca) [inline .github/workflows/ci-tests-app.yml:127]: You could remove the randomization from this workflow and get it merged for fabric and pl to unblock

File: .github/workflows/ci-tests-fabric.yml [hunk 7]
```diff
@@ -138,7 +138,7 @@ jobs:
       # NOTE: do not include coverage report here, see: https://github.com/nedbat/coveragepy/issues/1003
       run: |
         python -m coverage run --source ${{ env.COVERAGE_SCOPE }} \
-          -m pytest -v --timeout=30 --durations=50
+          -m pytest -v --timeout=30 --durations=50  --random-order-seed=42
```
Comment (carmocca) [inline .github/workflows/ci-tests-fabric.yml:141]: Shouldn't we use a random seed every time instead of a fixed one?

Otherwise there might still be interactions

Comment (Borda) [reply .github/workflows/ci-tests-fabric.yml:141]: @awaelchli had a good point with reproducible locally, so we can make see for a day. Would it work for you?

File: .github/workflows/ci-tests-app.yml (hunk 6 above)
Comment (Borda) [reply .github/workflows/ci-tests-app.yml:127]:
```suggestion
          -m pytest -m "not cloud" tests_app --timeout=300 -vvvv --durations=50
```

File: requirements/app/test.txt [hunk 8]
```diff
@@ -4,11 +4,13 @@ pytest-timeout ==2.1.0
 pytest-cov ==4.0.0
 pytest-doctestplus >=0.9.0
 pytest-asyncio ==0.20.3
+pytest-random-order ==1.1.0
```
Comment (Borda) [inline requirements/app/test.txt:7]:
```suggestion
# pytest-random-order ==1.1.0
```

File: .github/workflows/ci-tests-fabric.yml (hunk 7 above)
Comment (carmocca) [reply .github/workflows/ci-tests-fabric.yml:141]: I would assume that the seed used gets printed. Either way it's fine with me

**Review by carmocca — APPROVED** [review]

File: .github/workflows/ci-tests-fabric.yml (hunk 7 above)
Comment (Borda) [reply .github/workflows/ci-tests-fabric.yml:141]: ok, that would be also an option :)

**Review by awaelchli — APPROVED** [review]
